Add integration test to cover unsupported procols

brunoapimentel opened the pull request:
# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

brunoapimentel commented on tests/integration/test_yarn.py:
```diff
@@ -35,6 +35,20 @@
             marks=pytest.mark.xfail,  # temporary
             id="yarn_no_zero_installs",
         ),
```
I think that if we're 100% sure that `yarn info` won't trigger any arbitrary code execution, we can set `plugins=[]` after `yarn info` was called, and just before we do any `yarn install`. Remember that the user can define any foo plugin that can handle a foo protocol, so we'd need to account for that scenario.

By a quick read on the yarn source code that you linked, I'm not 100% sure that this is the case, it seems there's a lot going on there. I'm cloning the repo now so I can better look at it locally.

eskultety replied:
We discussed it some more over a private channel so I'm just summarizing our findings publicly for everyone to see:

- so far it looks like `plugins=[]` is really only for the user configured plugins, but not so much for the [default ones](https://v3.yarnpkg.com/features/plugins/#official-plugins) since even without this change the integration test provides expected results
- things change when deps include a custom locator based on a user-defined resolver plugin in which case there will be an error from Yarn about not being able to find a resolver
- parsing .yarnrc may be all we need to detect user-defined plugins (since it looks we can't affect the default ones much anyway) and hence produce a deterministic consistent error message by cachi2
- if we allow usage of plugins during the actual container build it will have consequences on the accuracy of the SBOM unless we make a note in there that a plugin (which we know nothing about) may have been executed during the build with unknown side-effects

brunoapimentel replied:
Ok, I found what the problem is. After this [PR](https://github.com/containerbuildsystem/cachi2/pull/399) came in, we started relying completely on the global cache. It seems that, after the dependency is cached, the original error no longer happens.

To reproduce it, try removing `~/.yarn/cache` and running cachi2 on the `no-available-resolver` branch on our test repo.

brunoapimentel replied:
It is still weird that calling `yarn info --all --recursive --cache` directly on the test repo does not cause the error. It is probably a combination of the yarn configuration we set on Cachi2. I'll try to figure this out tomorrow.

eskultety replied:
> To reproduce it, try removing ~/.yarn/cache and running cachi2 

Doh! Right you are! Yeah, now I'm seeing the error with both cachi2 and plain yarn.

> It is still weird that calling yarn info --all --recursive --cache directly on the test repo does not cause the error

Have you dropped the plugins from `.yarnrc` explicitly? Because I can reproduce both ways, even with plain `yarn info`.

In any case, based on my previous post, since the idea is to disable all plugins for now and knowing that Yarn v3 would fail spectacularly without declaring some of them, I can see us proceeding in either way:

- fetch the `plugins` setting from `.yarnrc` ([default](https://v3.yarnpkg.com/api/modules/plugin_compat.html) plugins are loaded implicitly) and if **any explicit** plugin is declared, fail with a nice error (e.g. `Plugins unsupported`) OR
- still fetch the `plugins` setting, but detect if `exec` is declared (in v3 `exec` is [contrib](https://v3.yarnpkg.com/features/plugins#contrib-plugins) has to be imported, [v4 loads it by default](https://yarnpkg.com/api/plugin-exec)) and if no other plugin is requested, make sure we add it to the list of plugins before running `yarn info` (we'll fail with an unsupported protocol error later anyway)

brunoapimentel replied:
Or maybe we can just remove explicit plugins from .yarnrc, since it'll allow `yarn info` to pass and the request will fail before `yarn install` is called. 

Thoughts on our options here, @taylormadore?

brunoapimentel replied:
So I discussed with @taylormadore and we decided on keeping the default plugins in yarnrc.

eskultety commented on cachi2/core/errors.py:
```diff
@@ -96,33 +96,22 @@ class FetchError(Cachi2Error):
 
-class GoModError(Cachi2Error):
-    """The 'go' command used while processing a Go module returned an error.
+class SubProcessError(Cachi2Error):
```
The name suggests it's a subclass of a subprocess error which it's not. Can we make it even more generic and call it perhaps `PackageManagerError`. We could then subclass this one further if needed for a specific package manager. The good thing about that would be that we could then choose during exception handling if we want to catch a specific error, or any package manager error or any Cachi2Error.

brunoapimentel replied:
Yeah, I made it SubProcess error in case we wanted to wrap other subprocesses in it (like Git or whatever else we end up calling). But I like the idea of `PackageManagerError`.

taylormadore replied:
I like PackageManagerError as well :+1:

brunoapimentel commented:
New push: 
- Dropped the commits that added a new error scenario for deps with no available resolver
- Added a commit to keep the default yarn plugins
- Renamed SubProcessError to PackageManagerError

eskultety commented on cachi2/core/package_managers/yarn/main.py:
```diff
@@ -109,6 +111,23 @@ def _configure_yarn_version(project: Project) -> None:
+    """Return a list of plugins that can be kept in .yarnrc.yml.
+
+    The plugins returned here are added by default by 'yarn init', or are included by default in
+    yarn 4+. They are not safe by default, but Cachi2 handles them in a safe way.
```
Also, we should make it more clear in the docstring that plugins returned by this helper are solely for means of being able to run `yarn info` successfully.

eskultety reviewed (commented):
> So I discussed with @taylormadore and we decided on keeping the default plugins in yarnrc.

@brunoapimentel  this alone may not be enough to fail the request properly with a consistent deterministic error though, will it? If a user declares their own plugin which happens to add a new protocol resolver, `yarn info` might still crash with that weird error which we're not handling anywhere in this revision. So unless I'm mistaken in my rationale we still need to either evaluate user's .yarnrc.yml somehow and detect a potentially problematic plugins (impossible unless we can somehow match the locator from package.json with the plugin) or just add back the yarn error handler you had in previous revisions (although a fragile solution).

eskultety commented on cachi2/core/package_managers/yarn/main.py:
```diff
@@ -109,6 +111,23 @@ def _configure_yarn_version(project: Project) -> None:
+    yarn 4+. They are not safe by default, but Cachi2 handles them in a safe way.
+    """
+    default_plugins = [
+        Plugin(
```
You can ignore ^this comment if you make the function docstring more explanatory, up to you, I'd be okay either way.

brunoapimentel replied:
I tried to address all of your outstanding comments by changing the function name, the docstring and also by only keeping the exec plugin in the whitelist.

The reasoning here is that we should only keep official plugins that implement a protocol (the others shouldn't affect the prefetch). It was also hard to figure out what exactly were the preloaded plugins in v4, and that list could change anyway :shrug:

brunoapimentel commented:
> > So I discussed with @taylormadore and we decided on keeping the default plugins in yarnrc.
> 
> @brunoapimentel this alone may not be enough to fail the request properly with a consistent deterministic error though, will it? If a user declares their own plugin which happens to add a new protocol resolver, `yarn info` might still crash with that weird error which we're not handling anywhere in this revision. So unless I'm mistaken in my rationale we still need to either evaluate user's .yarnrc.yml somehow and detect a potentially problematic plugins (impossible unless we can somehow match the locator from package.json with the plugin) or just add back the yarn error handler you had in previous revisions (although a fragile solution).

Right. I don't think it'll be easy to figure out if a plugin adds a locator, and if we fail in case of any non-default plugins, we might be blocking otherwise Cachi2-buildable repos (looking at [backstage](https://github.com/backstage/backstage/blob/master/.yarnrc.yml#L9)).

My reasoning here was that custom plugins+locators would be a rarer instance, and we could leave it for the user to figure out. But I agree that even that more fragile wrapper with a better error message is better than nothing, so I guess I'll add it back (good thing I didn't throw it away :rofl: )

taylormadore reviewed (approved):
Left a couple of minor comments/nitpicks, but overall LGTM :+1:
